enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,631 @@
+    if requirementHasMaxAge && !configHasMaxAge {
+        return fmt.Errorf("max age is required")
+    }
+    if requirement.MaxAge.LargestMaxAge != 0 && c.maxAge > requirement.MaxAge.LargestMaxAge {
+        return fmt.Errorf("does not match max age")
+    }
+    if requirement.MaxAge.SmallestMaxAge != 0 && c.maxAge < requirement.MaxAge.SmallestMaxAge {
+        return fmt.Errorf("does not match minimum age")
+    }
+
+    switch requirement.PreloadPolicy {
+        case DefaultPreloadPolicy, NoOpinionPreloadPolicy:
+        // anything is allowed, do nothing
+        case RequirePreloadPolicy:
+            if !c.preload {
+                return fmt.Errorf("preload must be true")
+            }
+        case RequireNoPreloadPolicy:
+            if c.preload {
+                return fmt.Errorf("preload must be unspecified")
+            }
+    }   
+
+    switch requirement.IncludeSubdomainsPolicy {
+        case DefaultIncludeSubdomains, NoOpinionIncludeSubdomains:
+        // anything is allowed, do nothing
+        case RequireIncludeSubdomains:
+            if !c.includeSubdomains {
+                return fmt.Errorf("includeSubdomains must be true")
+            }
+        case RequireNotIncludeSubdomains:
+            if c.includeSubdomains {
+                return fmt.Errorf("includeSubdomains must be unspecified")
+            }
+    }
+
+    return nil
+}
+// Check if the route matches the required domain/namespace in the HSTS Policy
+func requiredHSTSMatchesRoute(requirement RequiredHSTSPolicy, route *routeapi.Route, namespace *corev1.Namespace) (bool, error) {
+    matchesNamespace, err := matchesNamespaceSelector(requirement.NamespaceSelector, namespace)
+    if err != nil {
+        return false, err
+    }
+    if !matchesNamespace {
+        return false, nil
+    }
+    routeDomains := []string{route.Spec.Host}
+    for _, ingress := range route.Status.Ingress {
+        routeDomains = append(routeDomains, ingress.Host)
+    }
+    if matchesDomain(requirement.DomainPatterns, routeDomains) {
+        return true, nil
+    }
+
+    return false, nil
+}
+````
+
+Formatting of the HSTS Annotation field on a `route`:
+```yaml
+max-age=<numberofseconds>[;includeSubDomains][;preload]`
+```
+
+The `RequiredHSTSPolicies` variable can be configured by the administrator in the `Ingress`, as shown in this example, where
+there is one HSTS Policy is for domain abc.com, expires after one year, includes subdomains, and requests preload:
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: apps.abc.com
+  requiredHSTSPolicies:
+    - domainPatterns:
+        - abc.com
+      maxAge:
+          smallestMaxAge: 1
+          largestMaxAge: 31536000
+      preloadPolicy: "RequirePreload"
+      includeSubdomainsPolicy: "RequireIncludeSubdomains"
+```
+To ease administration, cluster administrators can use this simple command-line script, run with an input file of routes:
```

Comment:
Perhaps three levels of example:

To handle upgraded clusters with non-compliant HSTS routes already, the best solution is update the manifests at the source and apply the updates.  However, in upgrade clusters, you may wish to update the routes directly in the API.  There are a few common cases:

1. To apply HSTS to all routes in the cluster (after https://github.com/kubernetes/kubernetes/pull/101487 merges) , use `oc annotation -A haproxy.router.openshift.io/hsts_header=ax-age=31536000;preload;includeSubdomains`
2. To apply HSTS to all routes in a particular namespace, use  `oc annotation -nmy-namespace haproxy.router.openshift.io/hsts_header=ax-age=31536000;preload;includeSubdomains`
3. To apply more selective policy to running resources, a script like this can be used with an input file of routes: < your example here >

Reply:
Updated.